### DISCUSSION TITLE
On-Chain Gateway

### DISCUSSION THREAD

**Discussion by justinmoon** [fedimint/fedimint#3264] [General]
Our wallet module isn't currently suitable for heavy end-user usage because it can easily run out of UTXOs. If this happens, it doesn't just effect that one user but rather prevents ***all*** users of the federation from being able to withdraw on-chain until pending transactions are confirmed -- making more UTXOs available to the wallet module.

One mitigation is to improve the wallet module's UTXO management, but this is a long-term project.

Another option is to build an "on-chain gateway" like our current lightning gateway which would be able to process withdrawals (perhaps deposits, too?) for users without locking up the scarce resource of wallet module UTXOs.

**Comment by dpc** [comment]
Complexity level of: 

* improved UTXO mgmt: 5/10
* utxo gateway: `complexity_of(improved UTXO mgmt)` + 9/10

**Reply by elsirion** [reply]
I agree that the on-chain gateway idea is more complex, even when thinking of non-consensus code as much cheaper to change than consensus code (even though this consideration brings them closer together).

My biggest concern with relying on UTXO mgmt by the federation is that it's not adaptive. A professional gateway operator can easily adapt to changes in demand by adjusting fees, UTXO splitting behavior and also unstick transactions by paying higher fees out-of-pocket. Our federations simply cannot do this for lack of a on-chain fee pool (Liquid has this, but lacks a static guarantee never to run out afaik).

I do think we can make on-chain a fair bit more intelligent (e.g. through #3180 and also change splitting), but there will eventually be a UX ceiling that doesn't exist with the gateway approach. The question is if the ceiling will be high enough for us.

**Comment by EthnTuttle** [comment]
Could there be an optional on-chain escrow of sorts that users could opt-into or guardian could allocate a % of fund to? This escrow could be designed with being the first steps toward a UTXO Gateway.
